File: src/queries/api-keys.ts
/**
 * API key CRUD operations.
 *
 * API keys are stored in the `report` table with type = 'api-key'.
 * The `parameters` JSON field holds:
 *   - keyHash:   SHA-256 hex hash of the full key (mtk_...)
 *   - prefix:    first 8 chars of the key for display (mtk_xxxx)
 *   - scopes:    string[] of allowed scopes (currently: ["read", "write"])
 *   - createdAt: ISO timestamp of key creation
 *   - lastUsedAt: ISO timestamp of last use (nullable)
 *
 * The full key is only shown once at creation time.
 * We never store the raw key — only its hash.
 */
import { randomBytes, createHash } from "crypto";
import { rawQuery } from "@/lib/queries";

// ── Types ───────────────────────────────────────────────────────────────────

export interface ApiKeyRecord {
  id: string;
  name: string;
  prefix: string;
  scopes: string[];
  createdAt: string;
  lastUsedAt: string | null;
}

export interface ApiKeyCreateResult {
  id: string;
  name: string;
  key: string; // full key — shown only once
  prefix: string;
  scopes: string[];
  createdAt: string;
}

interface ReportRow {
  report_id: string;
  name: string;
  parameters: string | Record<string, unknown>;
}

// ── Constants ───────────────────────────────────────────────────────────────

const API_KEY_TYPE = "api-key";
// Use a fixed websiteId placeholder (report table requires it)
const PLACEHOLDER_WEBSITE_ID = "00000000-0000-0000-0000-000000000000";

// ── Helpers ─────────────────────────────────────────────────────────────────

function hashKey(key: string): string {
  return createHash("sha256").update(key).digest("hex");
}

function generateKey(): string {
  const bytes = randomBytes(32);
  return `mtk_${bytes.toString("base64url")}`;
}

function parseParams(params: string | Record<string, unknown>): Record<string, unknown> {
  if (typeof params === "string") return JSON.parse(params);
  return params;
}

// ── CRUD ────────────────────────────────────────────────────────────────────

/**
 * List all API keys for a user (without exposing the hash).
 */
export async function listApiKeys(userId: string): Promise<ApiKeyRecord[]> {
  const rows = await rawQuery<ReportRow>(
    `SELECT report_id, name, parameters
     FROM report
     WHERE user_id = {{userId::uuid}}
       AND type = '${API_KEY_TYPE}'
     ORDER BY created_at DESC`,
    { userId }
  );

  return rows.map((r) => {
    const p = parseParams(r.parameters);
    return {
      id: r.report_id,
      name: r.name,
      prefix: (p.prefix as string) || "mtk_???",
      scopes: (p.scopes as string[]) || ["read"],
      createdAt: (p.createdAt as string) || "",
      lastUsedAt: (p.lastUsedAt as string) || null,
    };
  });
}

/**
 * Create a new API key. Returns the full key (shown only once).
 */
export async function createApiKey(
  userId: string,
  name: string,
  scopes: string[] = ["read", "write"]
): Promise<ApiKeyCreateResult> {
  const key = generateKey();
  const keyHash = hashKey(key);
  const prefix = key.slice(0, 12) + "...";
  const now = new Date().toISOString();
  const id = crypto.randomUUID();

  const params = JSON.stringify({
    keyHash,
    prefix,
    scopes,
    createdAt: now,
    lastUsedAt: null,
  });

  await rawQuery(
    `INSERT INTO report (report_id, user_id, website_id, type, name, description, parameters)
     VALUES ({{id::uuid}}, {{userId::uuid}}, {{websiteId::uuid}}, {{type}}, {{name}}, {{description}}, {{params}}::jsonb)`,
    {
      id,
      userId,
      websiteId: PLACEHOLDER_WEBSITE_ID,
      type: API_KEY_TYPE,
      name,
      description: "",
      params,
    }
  );

  return { id, name, key, prefix, scopes, createdAt: now };
}

/**
 * Delete an API key by ID (only if owned by the user).
 */
export async function deleteApiKey(
  id: string,
  userId: string
): Promise<boolean> {
  const result = await rawQuery<{ count: bigint }>(
    `WITH deleted AS (
       DELETE FROM report
       WHERE report_id = {{id::uuid}}
         AND user_id = {{userId::uuid}}
         AND type = '${API_KEY_TYPE}'
       RETURNING 1
     )
     SELECT COUNT(*)::bigint AS count FROM deleted`,
    { id, userId }
  );
  return Number(result[0]?.count ?? 0) > 0;
}

/**
 * Validate an API key. Returns the userId if valid, null otherwise.
 * Also updates lastUsedAt.
 */
export async function validateApiKey(
  key: string
): Promise<{ userId: string; scopes: string[] } | null> {
  if (!key.startsWith("mtk_")) return null;

  const keyHash = hashKey(key);

  // Find the report row that contains this hash
  const rows = await rawQuery<{
    report_id: string;
    user_id: string;
    parameters: string | Record<string, unknown>;
  }>(
    `SELECT report_id, user_id, parameters
     FROM report
     WHERE type = '${API_KEY_TYPE}'`,
    {}
  );

  for (const row of rows) {
    const p = parseParams(row.parameters);
    if (p.keyHash === keyHash) {
      // Update lastUsedAt (fire and forget)
      const now = new Date().toISOString();
      const updatedParams = { ...p, lastUsedAt: now };
      rawQuery(
        `UPDATE report
         SET parameters = {{params}}::jsonb, updated_at = NOW()
         WHERE report_id = {{reportId::uuid}}`,
        { params: JSON.stringify(updatedParams), reportId: row.report_id }
      ).catch(() => {});

      return {
        userId: row.user_id,
        scopes: (p.scopes as string[]) || ["read"],
      };
    }
  }

  return null;
}

File: core/mantecato_core/queries/api_keys.py
"""
API key CRUD operations.

API keys are stored in the `report` table with type = 'api-key'.
The `parameters` JSONB column holds:
  - keyHash:   SHA-256 hex hash of the full key (mtk_...)
  - prefix:    first 12 chars of the key + "..." for display
  - scopes:    list of allowed scopes (e.g., ["read", "write"])
  - createdAt: ISO timestamp of key creation
  - lastUsedAt: ISO timestamp of last use (nullable)

The full key is only shown once at creation time.
We never store the raw key — only its hash.
SQL queries ported verbatim from TypeScript rawQuery calls.
"""

import base64
import hashlib
import json
import os
import uuid
from typing import Any

from mantecato_core.database import raw_query


API_KEY_TYPE = "api-key"
PLACEHOLDER_WEBSITE_ID = "00000000-0000-0000-0000-000000000000"


def hash_key(key: str) -> str:
    """SHA-256 hex hash of the API key."""
    return hashlib.sha256(key.encode()).hexdigest()


def generate_key() -> str:
    """Generate a new API key with mtk_ prefix."""
    random_bytes = os.urandom(32)
    return f"mtk_{base64.urlsafe_b64encode(random_bytes).decode().rstrip('=')}"


def _parse_params(params: Any) -> dict[str, Any]:
    """Parse parameters from either a string or dict."""
    if isinstance(params, str):
        return json.loads(params)
    return params or {}


async def list_api_keys(user_id: str) -> list[dict[str, Any]]:
    """List all API keys for a user (without exposing the hash)."""
    rows = await raw_query(
        """SELECT report_id, name, parameters
           FROM report
           WHERE user_id = {{userId::uuid}}
             AND type = 'api-key'
           ORDER BY created_at DESC""",
        {"userId": user_id},
    )

    result = []
    for r in rows:
        p = _parse_params(r.get("parameters"))
        result.append(
            {
                "id": r["report_id"],
                "name": r.get("name", ""),
                "prefix": p.get("prefix", "mtk_???"),
                "scopes": p.get("scopes", ["read"]),
                "createdAt": p.get("createdAt", ""),
                "lastUsedAt": p.get("lastUsedAt"),
            }
        )
    return result


async def create_api_key(
    user_id: str,
    name: str,
    scopes: list[str] | None = None,
) -> dict[str, Any]:
    """Create a new API key. Returns the full key (shown only once)."""
    if scopes is None:
        scopes = ["read", "write"]

    key = generate_key()
    key_hash = hash_key(key)
    prefix = key[:12] + "..."
    now = __import__("datetime").datetime.utcnow().isoformat()
    report_id = str(uuid.uuid4())

    params = json.dumps(
        {
            "keyHash": key_hash,
            "prefix": prefix,
            "scopes": scopes,
            "createdAt": now,
            "lastUsedAt": None,
        }
    )

    await raw_query(
        """INSERT INTO report (report_id, user_id, website_id, type, name, description, parameters)
           VALUES ({{id::uuid}}, {{userId::uuid}}, {{websiteId::uuid}}, {{type}}, {{name}}, {{description}}, {{params}}::jsonb)""",
        {
            "id": report_id,
            "userId": user_id,
            "websiteId": PLACEHOLDER_WEBSITE_ID,
            "type": API_KEY_TYPE,
            "name": name,
            "description": "",
            "params": params,
        },
    )

    return {
        "id": report_id,
        "name": name,
        "key": key,
        "prefix": prefix,
        "scopes": scopes,
        "createdAt": now,
    }


async def delete_api_key(key_id: str, user_id: str) -> bool:
    """Delete an API key by ID (only if owned by the user)."""
    rows = await raw_query(
        """WITH deleted AS (
             DELETE FROM report
             WHERE report_id = {{id::uuid}}
               AND user_id = {{userId::uuid}}
               AND type = 'api-key'
             RETURNING 1
           )
           SELECT COUNT(*)::bigint AS count FROM deleted""",
        {"id": key_id, "userId": user_id},
    )
    return int(rows[0].get("count", 0)) > 0 if rows else False


async def validate_api_key(key: str) -> dict[str, Any] | None:
    """
    Validate an API key. Returns {userId, scopes} if valid, None otherwise.
    Also updates lastUsedAt (fire and forget).
    """
    if not key.startswith("mtk_"):
        return None

    key_hash = hash_key(key)

    rows = await raw_query(
        """SELECT report_id, user_id, parameters
           FROM report
           WHERE type = 'api-key'
             AND parameters->>'keyHash' = {{keyHash}}""",
        {"keyHash": key_hash},
    )

    if not rows:
        return None

    row = rows[0]
    p = _parse_params(row.get("parameters"))

    # Update lastUsedAt (fire and forget — don't await)
    import datetime as _dt

    now = _dt.datetime.utcnow().isoformat()
    updated_params = {**p, "lastUsedAt": now}
    # Fire-and-forget: schedule but don't block
    import asyncio

    async def _update_last_used() -> None:
        await raw_query(
            """UPDATE report
               SET parameters = {{params}}::jsonb, updated_at = NOW()
               WHERE report_id = {{reportId::uuid}}""",
            {"params": json.dumps(updated_params), "reportId": row["report_id"]},
        )

    asyncio.ensure_future(_update_last_used())

    return {
        "userId": row["user_id"],
        "scopes": p.get("scopes", ["read"]),
    }
